### DISCUSSION TITLE
General Guidelines

### DISCUSSION THREAD

**Discussion by sigrunixia** [Obsidian-TTRPG-Community/TTRPG-Community-Admin#6] [The Burned Steak]
Continuing the Discussion from Discord, let's figure out what sort of standardization we want to go for. @Obsidian-TTRPG-Community/inquisitors 

Here are the Links @scottTomaszewski posted:
- https://docs.github.com/en/get-started/quickstart/github-flow
- https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/managing-protected-branches/about-protected-branches#require-pull-request-reviews-before-merging
- https://google.github.io/styleguide/

My Additions:
- https://help.obsidian.md/Contributing+to+Obsidian/Style+guide
- https://google.github.io/styleguide/docguide/style.html
- https://google.github.io/styleguide/docguide/philosophy.html

@Yilmas did not get a chance to really partake.

## Select Q&A

With a free Org account, we don't get standard code-owners as an across the org feature. I do believe it is still available per repo though as long as its a public one. 

So where do you want to start?

**Comment by sigrunixia** [comment]
https://www.conventionalcommits.org/en/v1.0.0/ is a part of [SEMVER](https://semver.org) spec. I think Conventional would be easier for _newbies_ than SEMVER pure. 
We do want to encourage this use out of the gate. 

My IDE which I commit to Git with helps me with these functions. I haven't checked if VSCode has similar.

**Reply by scottTomaszewski** [reply]
I'm down with it.  All the big IDEs should have some kind of helpers for commit formating. I use intellij and there is a plugin to validate commit formats with a regex and stuff.

**Reply by sigrunixia** [reply]
Oh which plugin (as that is what I use)?

Looks like VSCode has [VSCode Conventional Commits](https://github.com/vivaxy/vscode-conventional-commits).

**Reply by Yilmas** [reply]
I'm all for SEMVER, but I generally advise caution with most commit guides - including conventional.

**Reply by sigrunixia** [reply]
Could you detail your caution?

**Reply by Yilmas** [reply]
SEMVER is primarily a way of thinking, that not only benefits the dev or collaborators but also the end user (to some degree). Especially when it comes to release notes, where we can merge them - giving the end user a clear depiction of what has changed.

Conventional Commits however is a change in workflow. While it does come with benefits, mostly to collaborators, it is something better used in large organisations with multiple devs working across multiple projects. As it provides more transparency. On top of that, it often requires addons/extensions to work properly. 

- Additional overhead: Conventional Commits require developers to follow a specific format and syntax for their commit messages, which can add some overhead and cognitive load, especially for those who are not familiar with the convention.
- Potential resistance: Some developers may resist using Conventional Commits, either because they prefer a different format or because they perceive it as unnecessary or too rigid.
- **Risk of over-engineering**: While Conventional Commits can improve collaboration and maintainability, it's important to balance the benefits against the costs and avoid over-engineering or over-complicating the process.
- Learning curve: While the convention is straightforward, it may still take time for developers to get used to using it, which can slow down productivity in the short term.

From what I've seen on the discord, we are mostly dealing with people that want to help. But are not necessarily strong devs, with years of experience - rather they are just getting into Obsidian and often find even just Git daunting. The risk of over-engineering and its learning curve are too great in my opinion, short-mid term, versus the advantages and gains that can be achieved from it.

I would rather have us build up a good collection of collaborators, where quality of content takes precedence over quality of workflow.

Start small but concise. 

If we can get proper versioning in place, we will have won more than 50% of the battles just from that. Then as devs in each project start wanting more details related to a version, we can slowly introduce Conventional Commits. At that point we will have built a stronger community around those projects as well, where the evangelists will have more "power" to push things through.

It is similar to how I feel about lint. Great tool, if used in moderation. But if we just pull out all the stops now on lint, we will undoubtedly scare people away.

**Reply by sigrunixia** [reply]
That is very helpful and things I would have been unable to see. Javalent wants me to use conventional commits, so I have been. Mainly, because plugins are going to release-please style and are triggered automatically by those keywords. I would be lost without the plugins helping me though.

> Side note, this really is more of what you and Xentis want to do, I am more here to learn and be backup as needed.

I suppose versioning is where we start. :)

Edit: Is this [video simple enough](https://www.youtube.com/watch?v=wmw36DX7YEE) to include with general instructions? Mainly, most users just need to know major.minor.patch, right?

**Reply by Yilmas** [reply]
Javalent also isn't establishing something new, but rather maintaining a repo he "controls", and I doubt he will let anyone in without prior approval or experience. Where as the org and repos we are establishing here will generally come to rely on many people, both young and old, new and experienced. 
From a professional development viewpoint, I've used semver for ages, while commits are based on the individual developer. My philosophy is that it is more important that the dev explains their code through inline documentation, and normal docs, rather than the commits. (You only have a limited amount of hours a day to develope, that you need to prioritize what is important.)

I would love to hear what @scottTomaszewski thinks as well. While you (@sigrunixia) may be here "to learn and be a backup", you are still - in my book - the owner of this whole thing. So I'd rather have people pitch their thoughts, then just overruling stuff. 

Yes, that video explains it fine. Actually, what is important is the:
- Patch: Bugfixes
- Minor: Changes that are backward compatible.
- Major: Changes that are NOT backward compatible.

That is the essence of SEMVER. Which is why Chrome has received some "hate" over the years from the dev community, see image below and its major version.
![image](https://user-images.githubusercontent.com/6323003/236651834-925eceea-eb79-4faa-b6b0-79d1e6c0741a.png)

**Reply by sigrunixia** [reply]
I gave the team Admin access to the https://github.com/Obsidian-TTRPG-Community/TTRPG-Community-Admin Repo, so you will be able to easily add standards as they form.

**Reply by scottTomaszewski** [reply]
> Oh which plugin (as that is what I use)?
> 
> Looks like VSCode has [VSCode Conventional Commits](https://github.com/vivaxy/vscode-conventional-commits).

Sorry for the delay, "Git Toolbox" 
![image](https://github.com/Obsidian-TTRPG-Community/TTRPG-Community-Admin/assets/5295276/c1d3bb0e-a767-4076-9db5-065df01221ba)

Regarding my opinion - I generally agree that SEMVER is the most important thing overall.  Anything to help figure out whether the next release is major, minor, patch is something im willing to try.  Ive never been in an org that actually had automation tied to commit messages, but ive wanted to try it out.  I 100% hear you on the overhead.  My current dev team requires issue numbers as prefixes to the commit message and its pretty common to see a commit come in with message `ARCH-0000 Corrects typo` because who is going to open a ticket to correct a character in some documentation :sweat_smile: 

Totally agree on sigrunixia being a key decision maker and product owner :100: :100: :100:

**Reply by sigrunixia** [reply]
Fine fine, I'll be the fall guy. :P

So... how to best implement this when:
1. There are no actual releases for most projects as far as "builds". Are the commits
2. Across repos. I'm assuming we start every repo off at 0.0.0, no matter its content right now?
3. Defining the major, minor, and patch for our purposes. For example, later today I am going to pull apart the pf2e into its own repo. I would consider that.. major to the TTRPG share as it was a flagship part of it, but having some guidelines/examples are helpful.

**Reply by Yilmas** [reply]
According to the Semantic Versioning specification, a project or repo should start on version 0.1.0.

The major version 0 indicates that the project is in initial development and the API may be subject to breaking changes in the future. The minor version 1 indicates that there is already some functionality present and that the API may be extended in backward-compatible ways. Finally, the patch version 0 indicates that this is the initial release of the minor version.

Starting with version 0.0.0 is also a valid option, but it may not provide enough information to users about the stability and maturity of the project. Starting with version 0.0.1 is not recommended, as it implies that there was already a 0.0.0 version that was not released, which goes against the Semantic Versioning specification.

**Reply by sigrunixia** [reply]
0.1.0 Will work then I think. It gives us a starting point. 

Moving further into effectuating it: Do we want Dev branches?

**Reply by Yilmas** [reply]
Naaah, don't think that is necessary. Though potentially one could do a release branch.

**Reply by scottTomaszewski** [reply]
Some (GoLang) infer a `0` MAJOR version to be like the wild-west: you can add breaking changes and still call it a MINOR bump.  Essentially the code in question is considered alpha/beta, so people shouldnt rely on the structure or API yet.  

Once we go full `1.0.0` we should stick to semver strictly and consider what constitutes a breaking change.  I think the obvious case is renaming a file, the headers, blocks, etc will break people's links.  There is probably a good bit to think about here...

Sidenote on PATCH bumps - Ive seen a lot of developers get hung up on what counts (as a bugfix) here when it comes to breaking changes.  My general opinion is if a change corrects the code to conform to the public documentation (written or implied) then its a PATCH bump - even if its technically a breaking change.  For example, if the docs say "You can find a site map at note [[Site Index]]" and our site index note is called `Sote Index` - yes, technically a note rename is a "breaking" change, but the docs clearly imply that the existing note name is wrong.  So correcting the note name is only a PATCH release.  Im totally open to other ideas here, just offering a starting place :)

**Reply by sigrunixia** [reply]
That makes sense.

I think the next step is to start filling out the "Admin" repo with resources and best "practices policies". We'll keep it simple and build on it, but as was pointed out, we have a lot of new people and this can be intimidating.

**Reply by Yilmas** [reply]
> Some (GoLang) infer a `0` MAJOR version to be like the wild-west: you can add breaking changes and still call it a MINOR bump. Essentially the code in question is considered alpha/beta, so people shouldnt rely on the structure or API yet.

Funny you should mention it, this is another reason to dislike Star Citizen, I just got another email saying they are in "v3.19" which is an alpha. At that point, why even user versions when you f' it up so bad.

**Comment by sigrunixia** [comment]
Signing commits. Yay or nay?

**Reply by Yilmas** [reply]
Nay, too cumbersome. And it isn't like we are building some software that needs to be secure.

**Reply by sigrunixia** [reply]
Note and will update defaults :)

**Comment by sigrunixia** [comment]
We will need to make some guides/recommendations on setting up new repos and maintaining established repos. Here is what I currently have for the PF2E SRD branch protection. Note we do not have access to enterprise features.

Any suggested changes that would be the "default" and then tweaked based on each repos need?


<img width="914" alt="image" src="https://github.com/Obsidian-TTRPG-Community/TTRPG-Community-Admin/assets/68425372/f07d83f2-df38-4311-8c2a-028a42ac5bc7">

<img width="959" alt="image" src="https://github.com/Obsidian-TTRPG-Community/TTRPG-Community-Admin/assets/68425372/6f7e82a2-5282-4b83-a138-9e04ff972250">

<img width="934" alt="image" src="https://github.com/Obsidian-TTRPG-Community/TTRPG-Community-Admin/assets/68425372/03c10445-81d8-4177-8f83-dbe6065ec344">

**Reply by Yilmas** [reply]
Those are good :)

**Comment by sigrunixia** [comment]
A conversation on Squash, Merge, and rebase with Ebullient got me wondering if we need to, or want to, restrict what types of commits we'll do/accept. I know it's a contentious issue in some minds. https://www.atlassian.com/git/tutorials/merging-vs-rebasing Sold me on rebasing as far as GitHub is concerned, but I want to know your thoughts.

**Reply by Yilmas** [reply]
I'm a strong believer in commit history. Rebasing allows people to hide their mistakes, obscuring faults for the reviewers. 
However Interactive Rebasing done when performing local cleanups are a different matter.

At just about every job I've had since the conception of git, we've allowed merge on remote, and squash, rebase, etc. on local.

Or said in another way. What is the point of having a log, if you can rewrite the log at-will.

**Reply by scottTomaszewski** [reply]
I also really value commit history, mostly for the purpose of looking back at why/when things changed.  

In that sense, I'm totally fine with people doing a rebase/squash on their (remote) feature branches, but once a commit is pushed to a mainline (develop, main, master, etc), it's set in stone.  I'll actually encourage squashing a feature branch before merging into a mainline if it means others can look at the log and get a better idea of the feature change.  I don't particularly care that it took a dev 5 commits to find the correct DB port; squash that noise 😂

**Reply by sigrunixia** [reply]
Oki doki. The feature branch is probably something we ought to implement. However, it's not something we can force unless we pay for enterprise versions. So education is going to win here. 

Today I am going to be focused on filling out Admin repo with newbie getting started resources. A best/recommended practices for repo maintainers is probably warranted. Do either one of you have a few links that are go-to for that, which have not already been shared?

**Comment by sigrunixia** [comment]
Some quick updates:
- Resimplified the default labels. I'll leave the existing repos as they are for now. 
- Working on Org-Wide Templates (which will be simplistic).
- Learned real quick to not have multiple SRD focuses at once. D:
